### PR TITLE
feat(@desktop/general): Change user statuses

### PR DESCRIPTION
**Pull request by endulab**
Issue #5886

### What does the PR do

Make user statuses enum the same as in status-go.
User can choose 3 statuses (like in mobile): Always Online, Automatic, Inactive.
Handling updating own status on member list (Online/Inactive)

Always online - the status is broadcasted every 5 minutes. After finishing application user is visible as online for 2 weeks.
Automatic - the status is broadcasted every 5 minutes. After finishing application user is visible as Inactive.
Inactive - no broadcast, user is visible as Inactive :)

TODO:
Updating choose status menu to Figma design will be done in next task #6090

### Screenshot of functionality

Changing statuses:

https://user-images.githubusercontent.com/61889657/173034614-6773dc8a-28eb-4606-b02d-67e06d7b35a2.mov

Update after 5 minutes (Automatic status):

https://user-images.githubusercontent.com/61889657/173034706-d6f91a39-10d4-4b19-a722-df61841556b4.mov

### CODE REVIEW

**Review by elina2015 — APPROVED** [review]
Tested and Approved

File: src/app/modules/main/communities/module.nim [hunk 1]
```diff
@@ -105,7 +105,7 @@ method getCommunityItem(self: Module, c: CommunityDto): SectionItem =
           localNickname = contactDetails.details.localNickname,
           alias = contactDetails.details.alias,
           icon = contactDetails.icon,
-          onlineStatus = OnlineStatus.Offline, # TODO get the actual status?
+          onlineStatus = OnlineStatus.Inactive, # TODO get the actual status?
```
Comment (jrainville) [inline src/app/modules/main/communities/module.nim:108]: That's a leftover of me being lazy iirc lol.
I think we can just use `getStatusForContactWithId` from the contactService to get the status of the user.

File: src/app/modules/main/module.nim [hunk 2]
```diff
@@ -229,7 +230,7 @@ proc createChannelGroupItem[T](self: Module[T], c: ChannelGroupDto): SectionItem
         localNickname = contactDetails.details.localNickname,
         alias = contactDetails.details.alias,
         icon = contactDetails.icon,
-        onlineStatus = OnlineStatus.Offline,
+        onlineStatus = OnlineStatus.Inactive,
```
Comment (jrainville) [inline src/app/modules/main/module.nim:233]: Same as above, we can use `getStatusForContactWithId`.

File: src/app_service/service/settings/service.nim [hunk 3]
```diff
@@ -343,23 +343,17 @@ QtObject:
...
-      # set status to "offline"
-      if (self.saveSetting(KEY_SEND_STATUS_UPDATES, true)):
-        let r = status_update.setUserStatus(int(newStatus))
-        if(self.saveSetting(KEY_SEND_STATUS_UPDATES, value)):
-          self.settings.sendStatusUpdates = value
-          return true
+      const propagateStatus = true
+      if(self.saveSetting(KEY_SEND_STATUS_UPDATES, propagateStatus) != true):
+          return false
+      let _ = status_update.setUserStatus(newStatus.int)
```
Comment (jrainville) [inline src/app_service/service/settings/service.nim:351]:
```suggestion
      discard status_update.setUserStatus(newStatus.int)
```

**Review by jrainville — COMMENTED** [review]
Good work!

Found 2 small improvements and also a small oopsie

**Review by saledjenic — CHANGES_REQUESTED** [review]

File: src/app_service/common/conversion.nim [hunk 4]
```diff
@@ -63,3 +63,7 @@ proc wei2Eth*(input: string, decimals: int): string =
 
 proc wei2Gwei*(input: string): string =
   result = wei2Eth(input, 9)
+
+proc intToEnum*[T](intVal: int, defaultVal: T): T =
+  result = if (intVal >= ord(low(T)) or intVal <= ord(high(T))): T(intVal) else: defaultVal
```
[the remark this answers is not in the document]
Comment (endulab) [reply src/app_service/common/conversion.nim:68]: @saledjenic of course. I blindly copied this condition from dto/status_update.nim 🤦

[the remark this answers is not in the document]
Comment (endulab) [reply src/app_service/common/conversion.nim]: Fixed

File: src/app_service/service/contacts/service.nim [hunk 5]
```diff
@@ -38,8 +37,8 @@ type
 
 # Local Constants:
 const CheckStatusIntervalInMilliseconds = 5000 # 5 seconds, this is timeout how often do we check for user status.
-const OnlineLimitInSeconds = int(5.5 * 60) # 5.5 minutes
-const IdleLimitInSeconds = int(7 * 60) # 7 minutes
+const FiveMinsOnlineLimitInSeconds = int(5 * 60) # 5 minutes
+const TwoWeeksOnlineLimitInSeconds = int(14 * 24 * 60 * 60) # 2 weeks
```
Comment (osmaczko) [inline src/app_service/service/contacts/service.nim:41]: Just wondering why this is on the nim-client side instead of status-go? Ideally, we should just receive signal from status-go and update appropriate contact with the new online status.

File: src/app_service/service/settings/service.nim [hunk 6]
```diff
@@ -343,23 +343,17 @@ QtObject:
...
-      # The new user status needs to always be broadcast, so we need to update
-      # the settings accordingly and might turn it off afterwards (if user has 
-      # set status to "offline"
-      if (self.saveSetting(KEY_SEND_STATUS_UPDATES, true)):
-        let r = status_update.setUserStatus(int(newStatus))
-        if(self.saveSetting(KEY_SEND_STATUS_UPDATES, value)):
-          self.settings.sendStatusUpdates = value
-          return true
+      const propagateStatus = true
+      if(self.saveSetting(KEY_SEND_STATUS_UPDATES, propagateStatus) != true):
```
Comment (osmaczko) [inline src/app_service/service/settings/service.nim:349]: Why is that needed? In the current flow the same settings update may happen a few lines below :thinking:

**Review by osmaczko — APPROVED** [review]
Well done!

File: src/app_service/service/contacts/service.nim (hunk 5 above)
Comment (endulab) [reply src/app_service/service/contacts/service.nim:41]: It seems like a good idea.
Now, we are receiving signal from status-go every time the other peer broadcasts its status.
We keep all updated statuses and times of change in the cache and every 5 seconds we check the cache when the last status for every peer changed. Probably such a cache could be placed in status-go.
FYI @richard-ramos

File: src/app_service/service/settings/service.nim (hunk 6 above)
Comment (endulab) [reply src/app_service/service/settings/service.nim:349]: We need to set it 2 times. Before status_update.setUserStatus it needs to be set alway to true. W/o this the database will not be updated in status_update.setUserStatus and the status change will not be propagated to other peers. After status_update.setUserStatus we need to set it to true only in case of AlwaysOnline or Automatic in order to send 5 minutes broadcast info to the peers.

File: src/app_service/service/contacts/service.nim (hunk 5 above)
Comment (osmaczko) [reply src/app_service/service/contacts/service.nim:41]: > The 5s caching approach doesn't explain why nim client needs to check these intervals (5min and 2weeks) :thinking: Doesn't a status update come with online status already?

The only explanation I can think of is that this update is emitted only when peers broadcast their status over the network, which means we must check the interval to determine if they are offline or not. Nevertheless, I believe it should be done on status-go side, clients should only be interested in status updates, not the logic of how to obtain them. It also explains the "5s loop cache" mechanism, which is not really the optimization solution, it is rather more checking in the loop if contact has become offline or not.

**Review by jrainville — APPROVED** [review]

File: src/app/global/user_profile.nim [hunk 7]
```diff
@@ -15,8 +15,7 @@ QtObject:
     preferredName: string
     thumbnailImage: string
     largeImage: string
-    userStatus: bool
-    #currentUserStatus: int
+    currentUserStatus: int
```
[the remark this answers is not in the document]
Comment (endulab) [reply src/app/global/user_profile.nim:18]: I used int here because it is exposed to Qml. I was not sure if enum will work.

[the remark this answers is not in the document]
Comment (osmaczko) [reply src/app/global/user_profile.nim:18]: It's exposed through a property getter, so it could be of any type.

```diff
   proc getCurrentUserStatus*(self: UserProfile): bool {.slot.} =
-    self.currentUserStatus
+    self.currentUserStatus.int
```

File: src/app_service/service/settings/service.nim (hunk 6 above)
Comment (PascalPrecht) [reply src/app_service/service/settings/service.nim:349]: Correct. The comments in the removed code above try to elaborate on that.

@endulab might be worth keeping those comments as it otherwise causes confusion and we'll risk that someone will changes this code without knowing, which will cause regression (also in syncing)

**Review by PascalPrecht — COMMENTED** [review]
Changes look good to me so far. It was important for me to see whether the broadcast was still happening at all times (when setting the user status).

@endulab did you test this with multiple desktop instances and synced devices?

File: src/app_service/service/settings/service.nim (hunk 6 above)
Comment (endulab) [reply src/app_service/service/settings/service.nim:349]: @PascalPrecht you're right. I just saw that you added this comment and fixed setting status somewhere in the middle of this task. I removed this during syncing, but I will revert it.

File: src/app_service/service/contacts/service.nim (hunk 5 above)
Comment (endulab) [reply src/app_service/service/contacts/service.nim:41]: @osmaczko this is a good question. Regarding members/users the client is only interested in 2 states Online/Inactive. And I think the whole logic of checking statuses of members could be moved to status-go. At least, I do not see any problems with that. I will talk to Parvesh from mobile about a possibility of doing this.

**Comment by endulab** [comment]
> Changes look good to me so far. It was important for me to see whether the broadcast was still happening at all times (when setting the user status).
> 
> @endulab did you test this with multiple desktop instances and synced devices?

@PascalPrecht I tested desktop <-> desktop and desktop <-> mobile

**Comment by elina2015** [comment]
Tested and Approved
